File: src/main/java/de/danielmaile/aether/worldgen/ObjectManager.java
package de.danielmaile.aether.worldgen;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.stream.JsonReader;
import de.danielmaile.aether.Aether;
import de.danielmaile.aether.util.serializer.LocationSerializer;
import de.danielmaile.aether.util.serializer.VectorSerializer;
import de.danielmaile.aether.worldgen.dungeon.Dungeon;
import org.bukkit.Location;
import org.bukkit.util.Vector;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ObjectManager
{
    private static final String dungeonSavePath = Aether.getInstance().getDataFolder().getAbsolutePath()
            + File.separator + "data" + File.separator + "dungeons.aether";

    private List<Dungeon> dungeons = new ArrayList<>();

    public List<Dungeon> getDungeonList()
    {
        return dungeons;
    }

    private final Gson gson;

    public ObjectManager()
    {
        gson = new GsonBuilder()
                .registerTypeAdapter(Location.class, new LocationSerializer())
                .registerTypeAdapter(Vector.class, new VectorSerializer())
                .setPrettyPrinting()
                .create();
        loadDungeons();
    }

    public void save()
    {
        saveDungeons();
    }

    private void saveDungeons()
    {
        try (FileWriter fileWriter = new FileWriter(dungeonSavePath, false))
        {
            gson.toJson(dungeons, fileWriter);
            fileWriter.flush();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    private void loadDungeons()
    {
        try
        {
            JsonReader reader = new JsonReader(new FileReader(dungeonSavePath));
            dungeons = new ArrayList<>(Arrays.asList(gson.fromJson(reader, Dungeon[].class)));
        }
        catch (FileNotFoundException ignored) { }

        if (dungeons == null)
        {
            dungeons = new ArrayList<>();
        }
    }
}

File: src/main/java/de/danielmaile/aether/worldgen/ObjectManager.kt
package de.danielmaile.aether.worldgen

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.stream.JsonReader
import de.danielmaile.aether.Aether.Companion.instance
import de.danielmaile.aether.util.serializer.LocationSerializer
import de.danielmaile.aether.util.serializer.VectorSerializer
import de.danielmaile.aether.worldgen.dungeon.Dungeon
import org.bukkit.Location
import org.bukkit.util.Vector
import java.io.File
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.FileWriter
import java.util.*
import kotlin.collections.ArrayList


class ObjectManager {

    var dungeons = ArrayList<Dungeon>()

    private val dungeonSavePath = (instance.dataFolder.absolutePath
            + File.separator + "data" + File.separator + "dungeons.aether")
    private var gson: Gson = GsonBuilder()
        .registerTypeAdapter(Location::class.java, LocationSerializer())
        .registerTypeAdapter(Vector::class.java, VectorSerializer())
        .setPrettyPrinting()
        .create()

    init {
        loadDungeons()
    }

    fun save() {
        saveDungeons()
    }

    private fun saveDungeons() {
        if (dungeons.isEmpty()) return

        FileWriter(dungeonSavePath, false).use {
            gson.toJson(dungeons, it)
            it.flush()
        }
    }

    private fun loadDungeons() {
        try {
            val reader = JsonReader(FileReader(dungeonSavePath))
            dungeons = gson.fromJson<Array<Dungeon>>(reader, Array<Dungeon>::class.java).toCollection(ArrayList())
        } catch (ignored: FileNotFoundException) {
        }
    }
}